libsplinter/src/network/connection_manager/builder.rs

```diff
@@ -333,42 +337,42 @@ fn send_heartbeats<T: ConnectionMatrixLifeCycle, U: ConnectionMatrixSender>(
                 // if connection is already attempting reconnection, call reconnect
                 if reconnecting {
                     if last_connection_attempt.elapsed().as_secs() > retry_frequency {
-                        reconnections.push(endpoint.to_string());
+                        reconnections.push(metadata.clone());
                     }
                 } else {
-                    trace!("Sending heartbeat to {}", endpoint);
-                    if let Err(err) = matrix_sender
-                        .send(metadata.connection_id.clone(), heartbeat_message.clone())
+                    trace!("Sending heartbeat to {}", metadata.endpoint());
```

Comment:
Do we send a heartbeat to all connections, even if there is a duplicate endpoint?  If so, we should include the connection id as part of any connection logging here (e.g. `{endpoint} ({connection_id})`).  

As for my question, I suppose one of the connections could be closed remotely, but not the other, so we would have to detect that.